rp2040-hal/src/clocks/mod.rs

```diff
@@ -0,0 +1,213 @@
+//! Clocks (CLOCKS)
+// See [Chapter 2 Section 15](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
+
+use crate::pac::*;
+
+#[macro_use]
+mod macros;
+
+#[derive(Copy, Clone)]
+/// Provides refs to the CLOCKS block.
+pub struct ShareableClocks {
+    _internal: (),
+}
+
+impl ShareableClocks {
+    fn new(_clocks: &mut CLOCKS) -> Self {
+        ShareableClocks { _internal: () }
+    }
+
+    unsafe fn get(&self) -> &clocks::RegisterBlock {
+        &*CLOCKS::ptr()
+    }
+}
+
+/// Abstraction layer providing Clock Management.
+pub struct ClocksManager {
+    clocks: CLOCKS,
+    shared_clocks: ShareableClocks,
+}
+impl ClocksManager {
+    /// Exchanges CLOCKS block against Self.
+    pub fn new(mut clocks_block: CLOCKS) -> Self {
+        let shared_clocks = ShareableClocks::new(&mut clocks_block);
+        ClocksManager {
+            clocks: clocks_block,
+            shared_clocks: shared_clocks,
+        }
+    }
+
+    /// Releases the CLOCKS block
+    pub fn free(self) -> CLOCKS {
+        self.clocks
+    }
+
+    /// Getter for the Reference Clock.
+    pub fn ref_clock(&self) -> ReferenceClock {
+        ReferenceClock {
+            shared_dev: self.shared_clocks.clone(),
+        }
+    }
+
+    /// Getter for the System Clock
+    pub fn sys_clock(&self) -> SystemClock {
+        SystemClock {
+            shared_dev: self.shared_clocks.clone(),
+        }
+    }
+
+    /// Getter for the PeripheralClock
+    pub fn peripheral_clock(&self) -> PeripheralClock {
+        PeripheralClock {
+            shared_dev: self.shared_clocks.clone(),
+        }
+    }
+}
+
+/// For clocks with an integer divider.
+pub trait IntegerDivision {
+    /// Set integer divider value.
+    fn set_int_div(&mut self, div: usize);
+}
+
+/// For clocks with a fraction divider.
+pub trait FractionDivision {
+    /// Set fraction divider value.
+    fn set_frac_div(&mut self, div: usize);
+}
+
+/// For clocks that can have XOSC as source.
+pub trait XOSCClockSource {
+    /// Set XOSC as a source.
+    fn set_xosc_src(&mut self);
+}
+/// For clocks that can have ROSC as source.
+pub trait ROSCClockSource {
+    /// set ROSC as a source.
+    fn set_rosc_src(&mut self);
+}
+/// For clocks that can have ... itself (?) as a source (is that the "glitchless mux" ?)
```

Comment:
This is for when clocks have a number of sources where you can switch between them in a glitchless manner, and another set which _cannot_ be switched between without causing a clock glitch.

All the ones which will glitch are an "auxiliary source", and this tells the clock to do a glitchless switch between the current glitchless-capable clock and whatever is set as the auxiliary source.

A future enhancement that might want to have a helper: The idea is that when you want to switch between one of the clocks that can't be switched glitchless: you can switch the glitchless source to something else (such as the reference oscillator), switch the aux source to whatever new source you intend, and then switch the glitchless source back to the aux source.

@9names 
>Looks good. Not sure how I feel about setting a clock as it's own source though.
>I guess we'll find out how good an idea that is...

See above.

Reply:
Oh, this is supposed to be doing the glitchless aux switch?
Can I suggest it gets a different and more descriptive name if that's the case, because it just looks like all the other clk and aux selectors.

Reply:
It performs a glitchless switch to whatever the aux source/mux is set to.

Quoting the datasheet:
>All clock generators have a multiplexer referred to as the auxiliary (aux) mux. This mux has a conventional design
whose output will glitch when changing the select control. Two clock generators (clk_sys and clk_ref) have an additional multiplexer, referred to as the glitchless mux. The glitchless mux can switch between clock sources without generating a glitch on the output.

The functions ending in `_src` control that glitchless mux, the functions ending in `_auxsrc` control the auxiliary mux.

I have no objection to changing the name, but I'm not sure what would be better off the top of my head.

Reply:
Thanks for the explanation; I felt like the naming of the register's bits were quite confusing. What I understood was that those clocks had a source and an auxiliary source. But the naming of the bits sounds like the clock can be its own source, somehow ?! or its own auxiliary source. And so yeah, hence the `self` keyword and that comment !

Same here, no objection to changing the name but I'm still unsure what to call it exactly.

Reply:
Okay, so how about we talk about what we're trying to accomplish here - maybe the name will become obvious when we state what our goals are and how we are going to attempt to solve them.
And maybe it will become obvious if I'm missing a part of the problem...

Changing clock source (glitchless mux) is always glitchless.
Changing aux source (aux mux) can glitch, but that doesn't matter if it isn't selected as a clock source, 
So we don't want to switch aux when it's currently being used as the clock source.

1) Should we write this in a way that makes an aux-aux transition impossible while it's active?
2) Do we want a way of expressing that we want to switch the aux mux from one source to another, by switching clock source to something else (a fallback?), switching aux source, and then selecting aux as the clock source?

Reply:
We also have to consider the clocks without a glitchless mux. To properly switch the clocks with only an auxiliary source per the datasheet you need to disable the clock, wait 2 cycles of the source clock, switch the aux source, reenable the clock, and then finally wait an additional 2 cycles of the _new_ source clock.

If we are going to enforce the "switch glitchless mux away, switch aux mux, switch glitchless mux back" procedure, we should also enforce the "disable, switch aux mux, reenable" procedure.

The chip only has the glitchless mux for clocks that cannot be stopped.